replica/distributed_loader.cc

```diff
@@ -234,12 +234,13 @@ distributed_loader::get_sstables_from_upload_dir(distributed<replica::database>&
 }
 
 future<std::tuple<table_id, std::vector<std::vector<sstables::shared_sstable>>>>
-distributed_loader::get_sstables_from_object_store(distributed<replica::database>& db, sstring ks, sstring cf, std::vector<sstring> sstables, sstring endpoint, sstring bucket, sstring prefix, sstables::sstable_open_config cfg) {
-    return get_sstables_from(db, ks, cf, cfg, [bucket, endpoint, prefix, sstables=std::move(sstables)] (auto& global_table, auto& directory) {
+distributed_loader::get_sstables_from_object_store(distributed<replica::database>& db, sstring ks, sstring cf, std::vector<sstring> sstables, sstring endpoint, sstring bucket, sstring prefix, sstables::sstable_open_config cfg, std::function<seastar::abort_source*()> get_as) {
+    return get_sstables_from(db, ks, cf, cfg, [bucket, endpoint, prefix, sstables=std::move(sstables), &get_as] (auto& global_table, auto& directory) {
         return directory.start(global_table.as_sharded_parameter(),
-            sharded_parameter([bucket, endpoint, prefix] {
+            sharded_parameter([bucket, endpoint, prefix, &get_as] {
                 data_dictionary::storage_options opts;
-                opts.value = data_dictionary::storage_options::s3{bucket, endpoint, prefix};
+                seastar::abort_source* as = get_as ? get_as() : nullptr;
```

Comment:
can we default-initialize get_abort_source to a function returning nullptr?

Reply:
Seems more complicated than a conditional statement...

Reply:
but it's resolved in compile time and so saves the runtime condition

Reply:
On the other hand it has to appear in the header, so any changes "under the hood" has to escape the code unit. Four extra cycles in a relatively rare call seems like a very small price.

Reply:
Note also that since this is a std::function, having a "default" func instead would not remove any branches or anything, since invoke of std::func is wholly runtime.

Reply:
I think it may save a single condition since the default initialization of the parameter is determined at compile time rather than the above condition that happens at run time.  That said, it's a minute improvement, and this is not on the hot path, so I don't insist on it.

Reply:
No, because a default parameter is created on call site, so unless the whole method is inlined (which it is not, since it is in a different compilation unit than caller), the fact that method is none or not will not be known to the code here. 
Instead, however, here the inline of std::function will most likely be able to remove the check in the actual invoke, since we've already checked the method is callable -> bypass exception check.

Reply:
I presume that with default parameter initialization you can get rid of the above check for null function

Reply:
Again, default params are callsite created. The actual function code does not know anything about the actual param value (unless inlined), thus no checks can be eliminated by defaults. Only inlined and/or template functions can do that.

Reply:
Why do care to check if the std::function is default initialized if we provide an initializer for the function param?
There are no expected code paths that are expected to pass a null `std::function()`, or are there?